### PR TITLE
feat(select-v2): support allow-create feature

### PR DESCRIPTION
**Pull request by msidolphin**
support allow-create feature and refactored part of select-v2's code

Please make sure these boxes are checked before submitting your PR, thank you!

* [x] Make sure you follow Element's contributing guide [English](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.en-US.md) | ([中文](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.zh-CN.md) | [Español](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.es.md) | [Français](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.fr-FR.md)).
* [x] Make sure you are merging your commits to `dev` branch.
* [x] Add some descriptions and refer to relative issues for your PR.

### CODE REVIEW

File: packages/select-v2/src/useAllowCreate.ts [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+    return props.allowCreate && props.filterable
+  })
+
+  function hasExistingOption(query: string) {
+    if (!props.options) {
+      return false
+    }
+    return props.options.concat(states.createdOptions).some(option => {
+      return option.value === query
+    })
```
Comment (JeremyWuuuuu) [inline packages/select-v2/src/useAllowCreate.ts:20]: Can we test them individually instead of allocating new array and perform a `some` on it?

Comment (JeremyWuuuuu) [reply packages/select-v2/src/useAllowCreate.ts:20]: E.g.
```ts
hasValue = o => option.value === query

return props.options.some(hasValue) || states.createdOptions.some(hasValue)
```

File: packages/select-v2/src/useAllowCreate.ts [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+  function selectNewOption(option: Option) {
+    if (!enableAllowCreateMode.value) {
+      return
+    }
+    if (props.multiple && option.created) {
+      states.createdOptions.pop()
+      states.createdOptions.push({
+        ...option,
+        disabled: false,
+      })
```
Comment (JeremyWuuuuu) [inline packages/select-v2/src/useAllowCreate.ts:32]: Can you elaborate why popping before pushing?

File: packages/select-v2/src/useSelect.ts [hunk 3]
```diff
@@ -641,6 +650,13 @@ const useSelect = (props: ExtractPropTypes<typeof SelectProps>, emit) => {
     deep: true,
   })
 
+  // fix the problem that scrollTop is not reset in filterable mode
+  watch(filteredOptions, () => {
+    return nextTick(() => {
+      menuRef.value.resetScrollTop()
+    })
```
Comment (JeremyWuuuuu) [inline packages/select-v2/src/useSelect.ts:657]: Why would you return this anyway? Also, I think you can do:

```ts
nextTick(menuRef.value.resetScrollTop)
```

**Review by jw-foss — COMMENTED** [review]
I've left a few comments, please address it before this gets merged.

File: packages/select-v2/src/useAllowCreate.ts (hunk 2 above)
Comment (msidolphin) [reply packages/select-v2/src/useAllowCreate.ts]: sry, my mistake

**Comment by msidolphin** [comment]
> I've left a few comments, please address it before this gets merged.

done

**Review by jw-foss — APPROVED** [review]

**Pull request merged**